libbeat/autodiscover/providers/kubernetes/kubernetes.go

```diff
@@ -294,13 +296,12 @@ func NewLeaderElectionManager(
 		RetryPeriod:     2 * time.Second,
 		Callbacks: leaderelection.LeaderCallbacks{
 			OnStartedLeading: func(ctx context.Context) {
-				logger.Debugf("leader election lock GAINED, id %v", id)
-				eventID := fmt.Sprintf("%v-%v", metaUID, time.Now().UnixNano())
+				eventID = fmt.Sprintf("%v-%v", leaseId, time.Now().UnixNano())
```

Comment:
Why do we include the timestamp in the event ID? I'm wondering if we could drop it, and just use the leaseId? That way we don't need to store/reference `eventID`, and that would eliminate the question of data races entirely.

Reply:
I think it is risky and it could cause more problems. Example:

1. `metricbeat-1` is the leader.
2. There happens some kind of lease renewal.
3. `metricbeat-1` is the new leader again, but:
   1. `startLeading` from this new leader election is called.
   2. `stopLeading` is called after, triggered from the loss of lease from step 2. In that case, since they are using the same `eventId` we end up entirely without a leader because we deleted the configs.

If we use event id with timestamp, we make sure this never happens.

Reply:
Actually... we use `Run` for leader election:

And it says [this](https://github.com/kubernetes/client-go/blob/1518fca9f06c6a73fc091535b8966c71704e657b/tools/leaderelection/leaderelection.go#L204-L214):
> // Run starts the leader election loop. Run will not return
// before leader election loop is stopped by ctx or it has
// stopped holding the leader lease

This means we will never have the same leader twice! Because they stop running and they never are reelected!

So we have a problem. Example:
1. We have two nodes, `node-1` and `node-2`.
2. `node-1` is the first leader.
3. `node-1` loses the locker, so it stops running.
4. `node-2` gets elected.
5. There happens some kind of lease renewal that fails with timeout (for example, `rolebinding` gets deleted). `node-2` loses the lease. It stops running.
6. Who's going to be leader now? There are no more instances running to report the metrics...

So I believe the implementation as of now (the official one, not the from this branch) has two problems:

1. A metricbeat instance never stops from reporting metrics, even after losing the lease!
2. A metricbeat instance can never be reelected as the leader.

Reply:
To make it run indefinitely, we would have to put the code in this function in a for cycle:

https://github.com/elastic/beats/blob/4525d7867117d9f38b75e4fde41bfcf9895d95b5/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L330-L334

Maybe not ideal... Would it be just easier to use a lease and discard the leader election all together?

Reply:
I see two possible solutions:

1. Make leader election run again once it stops, so we know the metricbeat instance is always a candidate to be a leader.
   - Would this be a good idea? Maybe it is too resource consumption to be running this all the time, as the instance is constantly trying to acquire the lock.

2. Use a watcher that keeps track of the lease. This way, once the lease changes holder, we can start reporting metrics with this metricbeat instance.
   - Currently we don't have a watcher that tracks a single resource. This has been discussed before as a solution for the kubernetes secrets provider as well.
   - If we did this, we would be following the same logic for two providers, leader election and secrets.

Reply:
> This means we will never have the same leader twice! Because they stop running and they never are reelected!

Yikes... another good find.

I'm not familiar enough with Kubernetes leadership election or resource watchers, so it's probably best to discuss with others - but my instinct is to go with this one:

> Make leader election run again once it stops, so we know the metricbeat instance is always a candidate to be a leader.

I'm not sure if we can do better by using watchers, but this option is relatively small, so seems less risky to me. I don't think this would consume any more resources than having multiple Metricbeat instances that each attempt to acquire a lease?

In terms of code, I think we would need to change the `go le.Run(ctx)` to something like this:

```go
go func() {
    for {
        le.Run(ctx)
        select {
        case <-ctx.Done():
            return
        default:
            // Run returned because the lease was lost,
            // try to reacquire in case the new lease holder
            // loses the lease.
        }
    }
}()
```